pkg/storage/replica_test.go

```diff
@@ -4427,6 +4427,51 @@ func TestPushTxnUpgradeExistingTxn(t *testing.T) {
 	}
 }
 
+// TestPushTxnQueryPusheerHasNewerVersion verifies that PushTxn
+// returns pushee data in a request when a pusher is aware of a
+// newer version of the pushee.
+func TestPushTxnQueryPusheerHasNewerVersion(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	tc := testContext{}
+	stopper := stop.NewStopper()
+	defer stopper.Stop()
+	tc.Start(t, stopper)
+
+	key := roachpb.Key("key")
+	pushee := newTransaction("test", key, 1, enginepb.SERIALIZABLE, tc.Clock())
+	pushee.Priority = 1
+	pushee.Epoch = 12345
+	pushee.Sequence = 2
+	ts := tc.Clock().Now() // hlc.Timestamp{WallTime: 1}
+	pushee.Timestamp = ts
+	pushee.LastHeartbeat = &ts
+
+	pusher := newTransaction("test", key, 1, enginepb.SERIALIZABLE, tc.Clock())
+	pusher.Priority = 3
+
+	_, btH := beginTxnArgs(key, pushee)
+	put := putArgs(key, key)
+	if _, pErr := maybeWrapWithBeginTransaction(context.Background(), tc.Sender(), btH, &put); pErr != nil {
+		t.Fatal(pErr)
+	}
+
+	// Make the pushee in the request has updated information.
+	pushee.Priority++
+	args := pushTxnArgs(pusher, pushee, roachpb.PUSH_QUERY)
```

Comment:
Thanks! Tried that and found that `evalPushTxn` always calls `UpgradePriority` when a pusher wins..

```
	// Upgrade priority of pushed transaction to one less than pusher's.
	reply.PusheeTxn.UpgradePriority(args.PusherTxn.Priority - 1)
```

So, I tweaked the test a bit so that a push request will fail if `evalPushTxn` upgrades the pushee's priority to a higher value stored in the request. This might not be really realistic, though.